crates/iceberg/src/writer/file_writer/parquet_writer.rs

```diff
@@ -0,0 +1,594 @@
+        let to_write_null = RecordBatch::try_new(schema.clone(), vec![null_col]).unwrap();
+
+        // write data
+        let mut pw = ParquetWriterBuilder::new(
+            0,
+            WriterProperties::builder().build(),
+            to_write.schema(),
+            file_io.clone(),
+            loccation_gen,
+            file_name_gen,
+        )
+        .build()
+        .await?;
+        pw.write(&to_write).await?;
+        pw.write(&to_write_null).await?;
+        let res = pw.close().await?;
+        assert_eq!(res.len(), 1);
+        let data_file = res
+            .into_iter()
+            .next()
+            .unwrap()
+            // Put dummy field for build successfully.
+            .content(crate::spec::DataContentType::Data)
+            .partition(Struct::empty())
+            .build()
+            .unwrap();
+
+        // read the written file
+        let mut input_file = file_io
+            .new_input(data_file.file_path.clone())
+            .unwrap()
+            .reader()
+            .await
+            .unwrap();
+        let mut res = vec![];
+        let file_size = input_file.read_to_end(&mut res).await.unwrap();
+        let reader_builder = ParquetRecordBatchReaderBuilder::try_new(Bytes::from(res)).unwrap();
+        let metadata = reader_builder.metadata().clone();
+
+        // check data
+        let mut reader = reader_builder.build().unwrap();
+        let res = reader.next().unwrap().unwrap();
+        assert_eq!(to_write, res);
+        let res = reader.next().unwrap().unwrap();
+        assert_eq!(to_write_null, res);
+
+        // check metadata
+        assert_eq!(metadata.num_row_groups(), 1);
+        assert_eq!(metadata.row_group(0).num_columns(), 1);
+        assert_eq!(data_file.file_format, DataFileFormat::Parquet);
+        assert_eq!(
+            data_file.record_count,
+            metadata
+                .row_groups()
+                .iter()
+                .map(|group| group.num_rows())
+                .sum::<i64>() as u64
+        );
+        assert_eq!(data_file.file_size_in_bytes, file_size as u64);
+        assert_eq!(data_file.column_sizes.len(), 1);
+        assert_eq!(
+            *data_file.column_sizes.get(&0).unwrap(),
+            metadata.row_group(0).column(0).compressed_size() as u64
+        );
+        assert_eq!(data_file.value_counts.len(), 1);
+        assert_eq!(*data_file.value_counts.get(&0).unwrap(), 2048);
+        assert_eq!(data_file.null_value_counts.len(), 1);
+        assert_eq!(*data_file.null_value_counts.get(&0).unwrap(), 1024);
+        assert_eq!(data_file.key_metadata.len(), 0);
+        assert_eq!(data_file.split_offsets.len(), 1);
+        assert_eq!(
+            *data_file.split_offsets.first().unwrap(),
+            metadata.row_group(0).file_offset().unwrap()
+        );
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_parquet_writer_with_complex_schema() -> Result<()> {
```

Comment:
Thanks for adding this test! Can you make sure that you also assign the field-ids for the sub-fields?

```
parq /var/folders/22/yb9h2zd55ql37h4_50xkmg7r0000gn/T/.tmpAMDXHW/test-00000.parquet  --schema

 # Schema 
 <pyarrow._parquet.ParquetSchema object at 0x124b8f940>
required group field_id=-1 arrow_schema {
  optional int64 field_id=0 col0;
  optional group field_id=1 col1 {
    optional int64 field_id=-1 sub_col;
  }
  optional binary field_id=2 col2 (String);
  optional group field_id=3 col3 (List) {
    repeated group field_id=-1 list {
      optional int64 field_id=-1 item;
    }
  }
  optional group field_id=4 col4 {
    optional group field_id=-1 sub_col {
      optional int64 field_id=-1 sub_sub_col;
    }
  }
}
```

It would also be good to assert if they are there.

Reply:
> It would also be good to assert if they are there.

Sorry, I'm not sure what this means. Did you mean assert like following?
```
let schema = ...;
assert_eq(schema.fields[0].metadata().get(field_id),1);
```